### PR TITLE
Implement INTERSECT ALL, EXCEPT ALL

### PR DESCRIPTION
**Pull request by pranjalssh**
Implement INTERSECT ALL and EXCEPT ALL using window functions

Tracking Issue #15698

Using same approach as trinodb/trino#5890

```
== RELEASE NOTES ==

General Changes
* Implemented INTERSECT ALL
* Implemented EXCEPT ALL

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DistinctOutputQueryUtil.java [hunk 1]
```diff
@@ -106,7 +106,7 @@ public Boolean visitFilter(FilterNode node, Void context)
         @Override
         public Boolean visitIntersect(IntersectNode node, Void context)
         {
-            return true;
+            return node.isDistinct() || node.getSources().stream().anyMatch(source -> lookupFunction.apply(source).accept(this, null));
```
Comment (pranjalssh) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DistinctOutputQueryUtil.java:109]: Trino uses allMatch here, which I believe is incorrect(takes care of less cases)

**Comment by kaikalur** [comment]
I actually think our intersect and except implementations are not good. We should redo these to be better. One idea I'm thinking is to use row_number instead of count/gb which is quite expensive.

**Comment by kaikalur** [comment]
SELECT * FROM T1 INTERSECT SELECT * FROM T2

can be written as:

SELECT *, cardinality(SET_AGG(index) OVER(partition by <expand *>)) = 2 (or N if it's a N-way intersection)
(
SELECT * FROM (SELECT *, 1 as index, row_number() over(partition by <expand *>)  = 1 AS r FROM T1) WHERE r
UNION ALL
SELECT * FROM (SELECT *, 1 as index, row_number() over(partition by <expand *>)  = 1 AS r FROM T2) WHERE r
)

For INTERSECT ALL you don't need row_num = 1. Everything else remains the same.

**Comment by kaikalur** [comment]
SELECT * FROM T1 EXCEPT SELECT * FROM T2

can be written as:

SELECT FROM
(
SELECT *,
SET_AGG(index) OVER(partition by <expand *>) = ARRAY[1] 
AND index = 1 AS keep_in_result
(
SELECT * FROM (SELECT *, 1 as index, row_number() over(partition by <expand *>) = 1 AS r FROM T1) WHERE r
UNION ALL
SELECT * FROM (SELECT *, 2 as index, row_number() over(partition by <expand *>) = 1 AS r FROM T2) WHERE r
)
)
WHERE keep_in_result
For EXCEPT ALL you don't need row_num = 1 and index = 1 in the outer query. Everything else remains the same.

**Comment by kaikalur** [comment]
OTOH, looks like usage of these operators is limited so may not be worth the big change.

**Comment by kaikalur** [comment]
Also, this new idea will be most effective when we have row_number=1 pushdown to scan operator. So for now, don't change anything.

**Comment by yingsu00** [comment]
Adding some explanation to Sreeni's approach:

```
-- create a table
create table t1 (orderkey bigint, partkey bigint, suppkey bigint);
create table t2 (orderkey bigint, partkey bigint, suppkey bigint);
-- insert some values
INSERT INTO t1 VALUES (1, 2, 3), (1, 2, 3), (1, 2, 3), (2, 3, 4), (3, 4, 5), (3, 4, 5);
INSERT INTO t2 VALUES (1, 2, 3), (3, 4, 5), (4, 5, 6);

SELECT * FROM t1 EXCEPT  (select * from t2);

 orderkey | partkey | suppkey
----------+---------+---------
        2 |       3 |       4
```
This can be rewritten using window function as below:
```

WITH t1t2 as 
(
   SELECT
      orderkey,
      partkey,
      suppkey,
      1 as index,
      row_number() over(partition by orderkey, partkey, suppkey) as rowno 
   from t1 
   UNION all
   SELECT
      orderkey,
      partkey,
      suppkey,
      2 as index,
      row_number() over(partition by orderkey, partkey, suppkey) as rowno 
   from t2 
),
t1t2_withindex as 
(
   select
      set_agg(index) over(partition by orderkey, partkey, suppkey) as indexset,
      rowno,
      orderkey,
      partkey,
      suppkey 
   from t1t2 
)
select
   orderkey,
   partkey,
   suppkey 
from t1t2_withindex 
where
   indexset = array[1] 
   and rowno = 1;
```
However, `except all` CANNOT be rewritten simply by removing the filters from the above query. 

The correct answer for `SELECT * FROM t1 except all (select * from t2);` should be
```
 orderkey | partkey | suppkey
----------+---------+---------
        1 |       2 |       3
        1 |       2 |       3
        3 |       4 |       5
        2 |       3 |       4
```
If we just remove `indexset = array[1] and rowno = 1` we would output 3 (1,2,3)s and 2 (3,4,5)s. We need to add one more predicate that t1's row_number is greater than t2's rowcount. This requires a join and another `union all` as below

```
WITH t1t2 as (
  SELECT 
    orderkey, 
    partkey, 
    suppkey, 
    1 as index, 
    row_number() over(
      partition by orderkey, partkey, suppkey
    ) as rowno, 
    count(*) over(
      partition by orderkey, partkey, suppkey
    ) as rowcount 
  from 
    t1 
  UNION all 
  SELECT 
    orderkey, 
    partkey, 
    suppkey, 
    2 as index, 
    row_number() over(
      partition by orderkey, partkey, suppkey
    ) as rowno, 
    count(*) over(
      partition by orderkey, partkey, suppkey
    ) as rowcount 
  from 
    t2
), 
t1t2_withcount as (
  select 
    set_agg(index) over(
      partition by orderkey, partkey, suppkey
    ) as indexset, 
    index, 
    rowno, 
    rowcount, 
    orderkey, 
    partkey, 
    suppkey 
  from 
    t1t2
), 
t2_counts as (
  select 
    rowcount, 
    orderkey, 
    partkey, 
    suppkey 
  from 
    t1t2_withcount 
  where 
    index = 2
) 
select 
  t1t2_withcount.orderkey, 
  t1t2_withcount.partkey, 
  t1t2_withcount.suppkey 
from 
  t1t2_withcount 
  join t2_counts on t1t2_withcount.rowno > t2_counts.rowcount 
  and t1t2_withcount.orderkey = t2_counts.orderkey 
  and t1t2_withcount.partkey = t2_counts.partkey 
  and t1t2_withcount.suppkey = t2_counts.suppkey 
where 
  indexset <> array[2] 
  and index = 1 
  and t1t2_withcount.rowno > t2_counts.rowcount 
  and t1t2_withcount.orderkey = t2_counts.orderkey 
  and t1t2_withcount.partkey = t2_counts.partkey 
  and t1t2_withcount.suppkey = t2_counts.suppkey 
UNION ALL 
select 
  orderkey, 
  partkey, 
  suppkey 
from 
  t1t2_withcount 
where 
  indexset = array[1]
```
@kaikalur Sreeni, maybe my rewrite is not optimal. Do you have a better way to do it? By adding a join the performance might be just bad.

Anyways, we are going to implement these constructs more efficiently in Velox, so I'd say we just port over whatever PrestoSQL has for now.

**Pull request closed without merging**